### ISSUE TITLE
Wait for NickServ response before joining channels when nickserv_ident is used

### ISSUE THREAD

**Issue by osa1** [osa1/tiny#236]
We currently send NickServ identify message after RPL_WELCOME and join message after RPL_ENDOFMOTD. RPL_ENDOFMOTD is sent after RPL_WELCOME, but sometimes NickServ takes longer and authentication isn't successful until after RPL_ENDOFMOTD. When that happens we end up sending JOIN message before authenticating, which fails for channels with `+R` (login-to-join).

When nickserv_ident is used we should wait to hear from NickServ before sending JOIN.

Implementation is simple: just check for PRIVMSGs from NickServ in `update` in `libtiny_client/src/state.rs` and send join messages there. We'll need a `bool` field for whether we've sent JOIN already as we don't want to send JOIN every time we get a PRIVMSG from NickServ. The state should be `false` initially and reset on disconnect, and should be `true` after sending the JOIN message.

**Comment by osa1** [comment]
cc @trevarj , also ccing @Kabouik which I think mentioned this on the tiny IRC channel.

**Comment by trevarj** [comment]
I think this is a more reliable solution, because the PRIVMSGs from NickServ are unpredictable. I took the idea from this thread: https://github.com/sopel-irc/sopel/issues/310#issuecomment-22864583

This way, if the user configured nickserv_ident in their config, it will just retry joining failing channel until it connects. We could track retries, or just let it go forever... @osa1 
```
            // Reply 477 when user needs to be identified with NickServ to join a channel
            // Only if nickserv_ident is configured
            Reply { num: 477, params } => {
                if self.nickserv_ident.is_some() {
                    if let Some(chan) = params.get(1) {
                        debug!("Attempting to re-join channel {}", chan);
                        snd_irc_msg.try_send(wire::join(&[chan])).unwrap();
                    }
                }
            }
```

**Comment by osa1** [comment]
My suggestion is not to wait any specific message from NickServ, but wait for *anything*, i.e. any PRIVMSG from NickServ (whether it's rejection or confirmation) will cause the client to send JOINs. Would this still be more fragile than checking 477s?

My main concern with checking 477s is that there's still a race (though perhaps a benign one): sometimes we'll be able to authenticate before JOINs, sometimes we'll get one 477, sometimes more 477s. I think I'd prefer it if the client sent only one JOIN at the right time.

**Comment by osa1** [comment]
Note to self: #openjdk on OFTC is one of the channels that requires registration.

**Comment by trevarj** [comment]
I think that one timed JOIN after the first PRIVMSG from NickServ could still cause it to fail. Maybe I'm not following?

This is the log for attempting to join the channel after 477s.  
```
[libtiny_client/src/lib.rs:448] parsed msg: Msg { pfx: Some(Server("liquid.oftc.net")), cmd: Reply { num: 477, params: ["trevarj", "#vcs-home", "Cannot join channel (Need to be identified and verified to join this channel, \'/msg NickServ help\' to learn how to register and verify.)"] } }
[libtiny_client/src/state.rs:240] Attempting to re-join channel #vcs-home
first privmsg from NickServ (sending joins after this will fail still) ---> [libtiny_client/src/lib.rs:448] parsed msg: Msg { pfx: Some(User { nick: "NickServ", user: "services@services.oftc.net" }), cmd: PRIVMSG { target: User("trevarj"), msg: "This nickname is registered and protected.  If it is your nickname, you may", is_notice: true, ctcp: None } }
[libtiny_client/src/lib.rs:448] parsed msg: Msg { pfx: Some(User { nick: "NickServ", user: "services@services.oftc.net" }), cmd: PRIVMSG { target: User("trevarj"), msg: "authenticate yourself to services with the \u{2}IDENTIFY\u{2} command.  You are", is_notice: true, ctcp: None } }
[libtiny_client/src/lib.rs:448] parsed msg: Msg { pfx: Some(User { nick: "NickServ", user: "services@services.oftc.net" }), cmd: PRIVMSG { target: User("trevarj"), msg: "getting this message because you are not on the access list for the", is_notice: true, ctcp: None } }
[libtiny_client/src/lib.rs:448] parsed msg: Msg { pfx: Some(User { nick: "NickServ", user: "services@services.oftc.net" }), cmd: PRIVMSG { target: User("trevarj"), msg: "\u{2}trevarj\u{2} nickname.", is_notice: true, ctcp: None } }
[libtiny_client/src/lib.rs:448] parsed msg: Msg { pfx: Some(Server("liquid.oftc.net")), cmd: Reply { num: 477, params: ["trevarj", "#vcs-home", "Cannot join channel (Need to be identified and verified to join this channel, \'/msg NickServ help\' to learn how to register and verify.)"] } }
[libtiny_client/src/state.rs:240] Attempting to re-join channel #vcs-home
[libtiny_client/src/lib.rs:448] parsed msg: Msg { pfx: Some(Server("trevarj")), cmd: Other { cmd: "MODE", params: ["trevarj", "+R"] } }
[libtiny_client/src/lib.rs:448] parsed msg: Msg { pfx: Some(User { nick: "NickServ", user: "services@services.oftc.net" }), cmd: PRIVMSG { target: User("trevarj"), msg: "You are successfully identified as \u{2}trevarj\u{2}.", is_notice: true, ctcp: None } }
[libtiny_client/src/lib.rs:448] parsed msg: Msg { pfx: Some(User { nick: "trevarj", user: "~trevarj@109-252-138-192.dynamic.spd-mgts.ru" }), cmd: JOIN { chan: "#vcs-home" } }
```

**Comment by osa1** [comment]
Hmm, so the assumption here was that NickServ doesn't PRIVMSG to you for no reason, it only sends a message as a response to the messages you send. But as you showed this is not correct. Maybe handling 477 is the only reliable way then.

**Comment by osa1** [comment]
This was accidentally closed, #237 doesn't fix this. Reopening.
